File: LeetCodeProblemsGo/LeetCode429NaryTreeLevelOrderTraversal/LeetCode429NaryTreeLevelOrderTraversal.go
package LeetCode429NaryTreeLevelOrderTraversal

type Node struct {
	Val      int
	Children []*Node
}

func levelOrder(root *Node) [][]int {
	result := [][]int{}
	if root == nil {
		return result
	}
	currentLayer := []*Node{}
	nextLayer := []*Node{}
	currentLayer = append(currentLayer, root)
	for len(currentLayer) > 0 {
		layer := []int{}
		for len(currentLayer) > 0 {
			node := currentLayer[0]
			layer = append(layer, node.Val)
			currentLayer = currentLayer[1:]
			for _, child := range node.Children {
				nextLayer = append(nextLayer, child)
			}
		}
		result = append(result, layer)
		for len(nextLayer) > 0 {
			currentLayer = append(currentLayer, nextLayer[0])
			nextLayer = nextLayer[1:]
		}
	}
	return result
}

File: LeetCodeProblemsCMake/Problems/LeetCode429NaryTreeLevelOrderTraversal.cpp
//
// Created by saeed on 6/30/25.
//

#include "LeetCode429NaryTreeLevelOrderTraversal.h"

#include <queue>
using namespace std;
 vector< vector<int>> LeetCode429NaryTreeLevelOrderTraversal::levelOrder(Node *root) {
    vector<vector<int>> result{};
     if (root == nullptr)
         return result;
     queue<Node *> currentLayer{};
     queue<Node *> nextLayer{};
     currentLayer.push(root);
     while (!currentLayer.empty()) {
         vector<int> level{};
         while (!currentLayer.empty()) {
             auto node = currentLayer.front();
             currentLayer.pop();
             level.push_back(node->val);
             for (auto child : node->children) {
                 nextLayer.push(child);
             }
         }
         result.push_back(level);
         while (!nextLayer.empty()) {
             currentLayer.push(nextLayer.front());
             nextLayer.pop();
         }
     }
     return result;
}
